### PR TITLE
RFC: Add new factory data subsystem

### PR DESCRIPTION
**Pull request by rettichschnidi**
TODOs:
- [ ] Add usage documentation
- [x] Add usage example code
- [ ] Add (optional) integration into settings subsystem
- [x] Improve test coverage
- [ ] Improve test coverage to cover more (most) edge cases
- [ ] Discuss flash layout changes - how much pain does (would) it cause? Can we re-use the (no longer needed?) MCU Boot scratch partition?
- [x] Add Kconfig to enable factory data write API (disabled per default, enabled when selecting the factory data shell)
- [ ] Document code sizes for various implementations (dummy, LV, FCB without any other FCB users, FCB alongside a FCB backed settings module, EEPROM sample)
- [ ] Figure out proper locking strategy - mutexes needed at all?
- [ ] Reduce code duplication
- [ ] Use strnlen instead of strlen
- [ ] Figure out whether the user caller should supply the name-and-value-storing buffer instead of the factory data implementation allocating this (probably huge) buffer on the stack itself.

Factory data are device specific data like serial number, MAC addresses,
calibration data, and more. Factory data are expected to never be
changed (not even extended) during the whole (post-manufacturing)
lifetime of a device.

At first glance, the existing setting system seems like a solution to
this problem. However, (IoT) devices might use the settings module for
actual user settings. User settings data gets changed regularly,
contains personal and other relevant information like network keys and
therefore must be deleted (securely!) when a factory reset gets issued.

Non-volatile Storage (NVS), the recommended settings backend, does not
provide any functionality to deterministically, physically erase stored
data. Physical erasure of user data can only be achieved by deleting the
flash pages of the underlying NVS storage. Therefore, the same NVS
storage can not be used for factory data and user settings.

Moreover, while it is possible to register multiple source backends to
the settings module, there can be only one write backend at any time.
Also, support for multiple backends seems to be regarded as an anti
feature (see PR #19798), not worth to be pushed any further.

For those reasons, using the settings module is not an option.

This commit adds a new module named factory_data. Similar to the
settings module it allows to read and list entries. It will use a
different partition however and writing data is only possible when
CONFIG_FACTORY_DATA_WRITE is enabled. Except for manufacturing
firmwares, this should never be the case.

Unlike the settings module, this module uses the flash circular buffer
(FCB) storage per default, as NVS would waste one flash page for a
functionality this modules will never need: power-cut save writes.

A second, length-value based backend is available for devices with
limiting code size constraints, but reduced requirements regarding data
integrity.

A third, dummy backend is available to simplify code size measurements
by offering a baseline.

A user may also implement her own backend. An example for EEPROM is
provided.

This PR relates to #45211.

### CODE REVIEW

**Comment by Laczen** [comment]
@rettichschnidi, thank you for proposing the factory data system. It is a nice addition.

It does seem strange to use fcb for this, because:
a. fcb does not provide any way to describe the data, you are using a method similar to the settings system to describe the data (which is good, but does not reuse any of the code).
b. fcb is limited to flash only,
c. for systems that are using a different backend (nvs, fs) another backend will be added,
d. fcb is a circular buffer exactly to allow not loosing any data when power is interrupted, and as correctly stated this is not needed at all for the factory data application.

IMO it would be better to create some simple backend for the factory data subsystem that relies directly on flash (or flash_map) or another physical storage system (e.g. eeprom). The backend could be decoupled from the factory data system using simple read/write/erase/getsize routines, so that they could be replaced by something else (e.g. for eeprom usage). As the representation on the backend a simple TLV (type+length+value) or even LV (length+value) could be used that is ended by a `/0`. Walking over the data would be as simple as stepping over the length until the `/0` marker is found.

**Comment by rettichschnidi** [comment]
> a. fcb does not provide any way to describe the data, you are using a method similar to the settings system to describe the data (which is good, but does not reuse any of the code).

I tried to keep the code duplication low by using FCB instead of some (self written) solution (e.g. TLV). Yes, quite some code is very close to what settings does, but I could not find a good way to share code in a obvious, meaningful way. Could invest more time, likely come up with some improvements in this area.

b. fcb is limited to flash only

Thanks for mentioning this (again). Somehow I was not aware of this.

c. for systems that are using a different backend (nvs, fs) another backend will be added,

Would that be a problem? Seems the way the settings module has chosen and it seems a sensible design choice to me.

d. fcb is a circular buffer exactly to allow not loosing any data when power is interrupted, and as correctly stated this is not needed at all for the factory data application.

By not using a scratch partition the overhead of using FCB seem very low to me. At the same time, having data protected by checksums is something I'd love to have, yet not something I did not want to implement myself.

> IMO it would be better to create some simple backend for the factory data subsystem that relies directly on flash (or flash_map) or another physical storage system (e.g. eeprom). The backend could be decoupled from the factory data system using simple read/write/erase/getsize routines, so that they could be replaced by something else (e.g. for eeprom usage). As the representation on the backend a simple TLV (type+length+value) or even LV (length+value) could be used that is ended by a `/0`. Walking over the data would be as simple as stepping over the length until the `/0` marker is found.
> As an example of how to step in such a tlv system: https://github.com/Laczen/zephyrmodules/blob/master/zb8/zb8/src/zb8_tlv.c (In this example the tlv area is copied to ram before it is being read)

Actually, this look easy enough. Tacking on some CRC will probably not that complicated after all.

Bottom line: It looks like you have quite some track record in the NVS/Settings/etc. area. If my reasoning given above does not convince you, then the current FCB seems like a dead-end and I will come back with a new draft: TLV (with CRC), some kind of abstraction for read/write/erase/getsize routines

**Comment by Laczen** [comment]
@rettichschnidi, this is just my opinion. Other people might differ.

Adding a crc32 at the end of each record would be quite straightforward, the TLV would receive a slightly larger L value where the last 4 bytes would be the crc32 (this is also valid for the fcb data).

**Comment by rettichschnidi** [comment]
> Another (different approach) could be to store the data as CBOR objects, this could allow the factory data to be written ouside zephyr and directly copied to flash.

Just read this part. Might be even better than TLV.

**Comment by de-nordic** [comment]
If this is supposed to hold non-modifiable data, why not just put binary dump of structure at some specified address of flash or eeprom?
Using CBOR or any other FS will just add to flash, and if such APIs are not used elsewhere it is just a waste.
Dump of binary structure may be easily generated (serialized) and independently flashed (or written to external eeprom) independently from an application.

**Comment by rettichschnidi** [comment]
> If this is supposed to hold non-modifiable data, why not just put binary dump of structure at some specified address of flash or eeprom?

My (not previously expressed) motivation: Flexibility during production. Not necessarily having to release a new firmware when deciding to add a new field. For example, adding a new "manufactured timestamp" would be as easy as extending the Python script interacting with the device during production.

> Using CBOR or any other FS will just add to flash, and if such APIs are not used elsewhere it is just a waste.

I agree to some extend. However, I think the code size needed for reading out factory data can be kept reasonably low for most use cases. And if not, adding support for custom implementations surely would allow to address this point.

The code needed to write the factory data will never end up in a "live" firmware and therefore can be optimization for usability (whatever that is from a manufacturers POV).

> Dump of binary structure may be easily generated (serialized) and independently flashed (or written to external eeprom) independently from an application.

Another (unstated goal) of mine is to be able to store calibration data in the factory data partition. This data often can only be gathered when the Zephyr firmware is up and running. Therefore, upfront flashing is not a possibility. Shutting down Zephyr, flashing it, then restart and verify seems like a worse solution than to just spend some code on letting Zephyr write its own factory data.

**Comment by de-nordic** [comment]
> > If this is supposed to hold non-modifiable data, why not just put binary dump of structure at some specified address of flash or eeprom?
> 


> The code needed to write the factory data will never end up in a "live" firmware and therefore can be optimization for usability (whatever that is from a manufacturers POV).
> 
> > Dump of binary structure may be easily generated (serialized) and independently flashed (or written to external eeprom) independently from an application.
> 
> Another (unstated) goal of mine is to be able to store calibration data in the factory data partition. Some of this data can only be gathered when the Zephyr firmware is up and running. Therefore, upfront flashing is not a possibility. Shutting down Zephyr, flashing it, then restart and verify seems like a worse solution than to just spend some code on letting Zephyr write its own factory data.

A little bit in reverse order: got your point, I have gone with such process before where the device would be put into hw test/calibration/priming setup that would also upload the target firmware afterwards.
Because both firmware binaries would be based on the same source they would have the same, versioned,  hw description structure; the only difference would be that the application firmware would be able to read and process 2-3 versions back, to make it work with older hardware (when possible).
Nevertheless the advantage of this approach, when the structure was stored within soc flash, was that it was addressed as ordinary structure that has been stored by test/cal/prim firmware at specified flash location and not overwritten by uploading of the application firmware. Of course such structure is aligned to some page address and you have to have some CRC and version field to verify that it is the structure you expect.
Additional feature of such approach is that you could also re-create the hardware description, while servicing the board, from your factory db and generate hex to flash into the board, when calibration was not required, shortening the process - this is specifically useful when you have separate sensor board and computation board and need only to replace the later, while having the former already calibrated within previous set.

> I agree to some extend. However, I think the code size needed for reading out factory data can be kept reasonably low for most use cases. And if not, adding support for custom implementations surely would allow to address this point.
> 

You can not guarantee that. mcuboot is an example of this: however you will try to make mcuboot small and efficient, you are dependent on Zephyr as the base system for it, so any significant increase in codes size in areas touching flash access (API, bus API and so on) within Zephyr, will increase (or decrease) the flash size requirements within mcuboot, with the later having no control over that happening.

> My (not previously expressed) motivation: Flexibility during production. Not necessarily having to release a new firmware when deciding to add a new field. For example, adding a new "manufactured timestamp" would be as easy as extending the Python script interacting with the device during production.

The flexibility, as far as I can tell from my experience, considering the fact that you do not modify the data in-field, works only one way: you have to support older hardware that is in-field, with old calibration structures, with a new firmware, unless you retrieve these devices and put them through the factory process again.
When you release new hardware, or hardware with the new fields in flash, you should already have new structures describing them, even if your firmware may lag behind and not be able to use them. Nevertheless your firmware should be able to recognize, and process, previous versions of structures at lease a few versions back.

**Comment by rettichschnidi** [comment]
@de-nordic Thanks for the feedback and sorry for the long delay.

> A little bit in reverse order: got your point, I have gone with such process before where the device would be put into hw test/calibration/priming setup that would also upload the target firmware afterwards. Because both firmware binaries would be based on the same source they would have the same, versioned, hw description structure; the only difference would be that the application firmware would be able to read and process 2-3 versions back, to make it work with older hardware (when possible). 

Surely possible, but it seems rather inflexible. I think it is a valuable to a) be able to add arbitrary new data fields at the production site and b) to access the factory data using a well defined/know mechanism. Hence the "string-key" to value interface, which is similar to the the settings system.

Having to create and deploy a new factory firmware for even the slightest change is something I'd really want to avoid. Also, I think having to know and support every past data structure is (at least) as complicated as using the "string-key"-value interface and assigning sensible default values on the caller site when a value does not exist.

> Nevertheless the advantage of this approach, when the structure was stored within soc flash, was that it was addressed as ordinary structure that has been stored by test/cal/prim firmware at specified flash location and not overwritten by uploading of the application firmware.

How is this a unique feature compared to my proposed factory data? The factory data (FCB or any other) would also be stored in a dedicated partition on a well defined address.

> Of course such structure is aligned to some page address and you have to have some CRC and version field to verify that it is the structure you expect.

FCB does exactly this. One CRC per entry and not a single one over all of the data, but I think it is sensible to reuse what is already existing.

> Additional feature of such approach is that you could also re-create the hardware description, while servicing the board, from your factory db and generate hex to flash into the board, when calibration was not required, shortening the process - this is specifically useful when you have separate sensor board and computation board and need only to replace the later, while having the former already calibrated within previous set.

This should work with every kind of storage backend. Be it FCB, TLV or just a structure dump.

> You can not guarantee that. mcuboot is an example of this: however you will try to make mcuboot small and efficient, you are dependent on Zephyr as the base system for it, so any significant increase in codes size in areas touching flash access (API, bus API and so on) within Zephyr, will increase (or decrease) the flash size requirements within mcuboot, with the later having no control over that happening.

I did not think about MCUBoot. However, a bootloader should be as simple as possible. Why would it need to access any factory data?

> The flexibility, as far as I can tell from my experience, considering the fact that you do not modify the data in-field, works only one way: you have to support older hardware that is in-field, with old calibration structures, with a new firmware, unless you retrieve these devices and put them through the factory process again.

This is my intention. And by not having fixed structures, new FW versions on older HW can deal with `factory_data_load_one()` returning -ENOENT by just using a sensible default value.

> When you release new hardware, or hardware with the new fields in flash, you should already have new structures describing them, even if your firmware may lag behind and not be able to use them.

Not needed when having a (more flexible) key-value factory data API.

> Nevertheless your firmware should be able to recognize, and process, previous versions of structures at lease a few versions back.

In my case: Every single version ever produced :)

Having said all of above and worked a bit on a new PR revision:
* Would you agree on having a key-value API, similar to the settings module, is a good way to go?
* I am tempted to implement the current key-value API by actually using a struct as backend. Just to prove that it is flexible enough to allow this too. My hope is that this will result in a small code size. Thoughts?
* Ideally, I can write up a benchmark of the FCB vs TLV vs custom struct implementations. What else other than code size would be important?
* I realized that I'd rather not put a IO abstraction layer behind the API anymore and save the code size overhead. Will instead make the reusable parts accessible for custom factory data implementations.

**Comment by rettichschnidi** [comment]
@de-nordic @Laczen Just pushed a new revision and updated the original issue on the top. While I have not everything (yet) addressed, I'd love to hear whether you think this is going in the right direction or not.

**Comment by Laczen** [comment]
@rettichschnidi, this seems to be getting very complex and maintaining this might get problematic. I am not a fan of the different solutions for the different backends. This is just my opinion.

**Comment by carlescufi** [comment]
@rettichschnidi is this something you'd like to bring to the Architecture WG?

**Comment by rettichschnidi** [comment]
> @rettichschnidi is this something you'd like to bring to the Architecture WG?

This PR is surely something I intend to push forward one way or another. We are using this already internally at my employer.

Would it be OK to bring it to the WG even in its current state (quite some open TODO)? If so I'd be happy to this soon-ish, otherwise I'd need at least another 2-3 weeks to address (the most glaring of) the open TODOs.

File: subsys/factory_data/factory_data_fcb.c [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+	}
+
+	/* Missing strnlen, but input got written by us, guaranteed to have have a trailing '\0' */
+	name_len = strlen((const char *)buf);
+	if (name_len == 0) {
+		/*  */
+		__ASSERT(false, "Zero length names are not allowed");
+		return -EIO;
+	}
+	return memcmp(name, buf, name_len) == 0; /* positive return value on match */
```
Comment (mlasch) [inline subsys/factory_data/factory_data_fcb.c:107]: This does not allow having entries with the same prefixes, like `test` and then adding `test2` etc. I think here we should use `strncmp` or similar.

**Comment by rettichschnidi** [comment]
Will reopen once rebased to Zephyr 3.2.
